### ISSUE TITLE
How to avoid failed actions because of unable to update status to "in-progress"?

### ISSUE THREAD

**Issue by Dekadinious** [Prospress/action-scheduler#1053]
I recently developed a system that uses ActionScheduler to schedule a large number of async actions that are critical that they actually run. I have never seen this before, but from a loop of 5336 added async actions, 28 of them failed with:

`action failed via Async Request: Unable to update the status of action 18605640 to in-progress.`

When can this happen and how can I mitigate it?

How can I subscribe to updates when this happens to actions so that I can alert my team?

**Comment by barryhughes** [comment]
Hi @Dekadinious,

To observe and potentially take some sort of follow-up action, you might hook into `action_scheduler_failed_execution` (which is what the logger uses to write the message you quoted):

```php
/**
 * Runs when action execution fails.
 *
 * @param int       $action_id Action ID.
 * @param Exception $e         Exception instance.
 * @param string    $context   Execution context.
 */
do_action( 'action_scheduler_failed_execution', $action_id, $e, $context );
```

> When can this happen and how can I mitigate it?

Updating the status to `in-progress` happens after a batch of actions has been claimed by the queue processor. 

Each action is individually processed, and the status is changed to make it clear to other processes that work is underway (important sidebar for us to look into later: there is a probable race condition within the corresponding check inside [`ActionScheduler_Abstract_QueueRunner::process_action`](https://github.com/woocommerce/action-scheduler/blob/3.7.4/classes/abstracts/ActionScheduler_Abstract_QueueRunner.php#L78-L81)).

Why might it fail? Lots of reasons. Possibilities include:

- A race condition between competing queue runner processes (despite the above note, I'd hope this is generally not happening, because there are other safeguards in place).
- The SQL used to make the update may be broken (because it passes through `wpdb`, it is filterable).
- The query may simply fail for reasons outside the direct control of Action Scheduler _(db gone away,_ etc).

It's hard to suggest a general mitigation—though a deferred retry attempt is one possibility—but we're definitely open to ideas here.

**Comment by barryhughes** [comment]
I'm going to close the issue, since it's not clear we need to make any changes (of course, let me know if you feel we should re-open).

**Comment by Dekadinious** [comment]
@barryhughes Sorry for answering this so late. Hooking into the failed action and doing a retry would seem like something that can work in my case :)

**Comment by barryhughes** [comment]
Great! We've also tried to make some of these errors/exception messages a little more informative, and we're thinking about ways to remodel error/exception handling in general, but there's quite a bit more work to do there.

**Comment by JUVOJustin** [comment]
I just wanted to report that I am also running into this frequently. I build a wrapper that heavily utilizes action scheduler to mass import thousands of products. We use up to 8 queues in parallel each having a batch size of 2. execution is done via WP Cli. 

The 3-4000 actions sync takes around 7 minutes and runs every two hours. Nearly every seconds sync has one action failing due to the failed status update. 

Is this still being investigated? We took immense efforts to correctly lock data access in our wrapper to allow a lot of parallel executions. I would love to solve this remaining puzzle piece as well. Let me know if I can help.

**Comment by Dekadinious** [comment]
@JUVOJustin FWIW: We also have a product sync from an external system. We have about 30 000 SKUs that are imported. We just coded our own thing directly and also use WP CLI to run it. Much more stable because you can error handle directly and know if stuff runs or not. But these sometimes failing AS tasks can still impact a lot, so it would be great to have it solved.

**Comment by barryhughes** [comment]
> Is this still being investigated?

No, this issue was closed. There is always going to be some potential for this sort of problem, especially under heavy load. Additionally, the system "self-recovers" from problems like this—at least in some senses:

- If the status could not be updated because of a problem at MySQL-level, then it will remain *'pending'*. The queue cleaner will subsequently release it from its claim, and it can then be picked up again by a future queue runner.
- Or, if our locks failed and a parallel worker picked up the action, then it will presumably be executed via that other process in any case (despite an error being raised in the current process).

> We use up to 8 queues in parallel each having a batch size of 2.

The right strategy obviously depends on your server resources and the nature of the workload being processed, but I wonder if the high workload and low batch size is causing some degree of 'thrashing'. If the tasks are generally of a very similar nature, and take roughly the same amount of time to run, perhaps it would be advantageous to increase the batch size?

If you'd like to raise a fresh issue to discuss this more, or feel out possible mitigations from within Action Scheduler, the team would definitely be happy to discuss it :-)
